### PR TITLE
levenshtein: fixup Csharp

### PR DESCRIPTION
**Pull request by AhmedZero**
1- Replaced `string` with `ReadOnlySpan<char>` in Levenshtein function to avoid additional allocations and improve efficiency for slicing strings.
2- Use stackalloc instead of new.

### CODE REVIEW

**Comment by Ichoran** [comment]
This looks good to me--much more like the reference implementation!

**Comment by Akeit0** [comment]
You can add ``[MethodImpl(MethodImplOptions.AggressiveOptimization)]`` to levenshtein,
And you can use swapping instead of coping.
```cs
 // Swap spans
var temp = prev;
prev = curr;
curr = temp;
```

File: levenshtein/csharp/code.cs [hunk 1]
```diff
@@ -18,32 +18,60 @@
...
+        prev[i] = i;
     }
 
-    for (int i = 1; i <= m; i++)
+    // Iterate and compute distance
+    for (int i = 1; i <= n; i++)
     {
-        for (int j = 1; j <= n; j++)
+        curr[0] = i;
+        for (int j = 1; j <= m; j++)
```
Comment (Akeit0) [inline levenshtein/csharp/code.cs:60]: ``j <= m`` should be replaced with ``j<curr.Length``.
This would reduce boundary checks.

**Comment by AhmedZero** [comment]
> You can add `[MethodImpl(MethodImplOptions.AggressiveOptimization)]` to levenshtein, And you can use swapping instead of coping.
> 
> ```cs
>  // Swap spans
> var temp = prev;
> prev = curr;
> curr = temp;
> ```
I use Disasmo.
the `[MethodImpl(MethodImplOptions.AggressiveOptimization)]` is not effective.

**Comment by AhmedZero** [comment]
@PEZ why is there no review?

**Comment by PEZ** [comment]
@AhmedZero I'm sorry for the radio silence. We're doing this as a side project in addition to working for food and taking care of our families. 😀

I'll try to give your changes a look tomorrow. 

I'm on my phone and can't check right now, but if you haven't attached benchmark results from before and after your changes, please do that. It makes it much easier for us to assess things.

**Comment by AhmedZero** [comment]
Old version:
Output:
```
Checking C#
Check passed
Benchmarking C#
Benchmark 1: ./csharp/code/code aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa ...
  Time (mean ± σ):      82.9 ms ±   1.3 ms    [User: 72.9 ms, System: 7.6 ms]
  Range (min … max):    81.5 ms …  83.9 ms    3 runs
```

New Version:
```
Checking C#
Check passed
Benchmarking C#
Benchmark 1: ./csharp/code/code aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa ...
  Time (mean ± σ):      52.0 ms ±   1.7 ms    [User: 40.1 ms, System: 10.5 ms]
  Range (min … max):    50.3 ms …  53.8 ms    3 runs
```

new version is approximately 37.27% faster than the old version.

**Comment by PEZ** [comment]
🎉 On my machine this also fixes so that I can actually run the executable. The previous code resulted in some index out of bounds and then just hanged.

Also about that performance gain. It is probably much more than 37%. Since we have the CLR start time in the mix. I'm now very curious about the AOT build. I fail to compile it on my machine:

```
dotnet publish csharp -o csharp/code-aot /p:PublishAot=true /p:OptimizationPreference=Speed
Restore complete (0.5s)
  code failed with 1 error(s) (4.7s) → csharp/bin/Release/net9.0/osx-arm64/code.dll
    /Users/pez/.nuget/packages/microsoft.dotnet.ilcompiler/9.0.0/build/Microsoft.NETCore.Native.targets(376,5): error MSB3073: The command ""clang" -ld_classic "obj/Release/net9.0/osx-arm64/native/code.o" -o "bin/Release/net9.0/osx-arm64/native/code" -exported_symbols_list "obj/Release/net9.0/osx-arm64/native/code.exports" -Wl,-dead_strip -gz=zlib /opt/homebrew/Cellar/dotnet/9.0.101/libexec/packs/runtime.osx-arm64.Microsoft.DotNet.ILCompiler/9.0.0/sdk/libbootstrapper.o /opt/homebrew/Cellar/dotnet/9.0.101/libexec/packs/runtime.osx-arm64.Microsoft.DotNet.ILCompiler/9.0.0/sdk/libRuntime.WorkstationGC.a /opt/homebrew/Cellar/dotnet/9.0.101/libexec/packs/runtime.osx-arm64.Microsoft.DotNet.ILCompiler/9.0.0/sdk/libeventpipe-disabled.a /opt/homebrew/Cellar/dotnet/9.0.101/libexec/packs/runtime.osx-arm64.Microsoft.DotNet.ILCompiler/9.0.0/sdk/libstandalonegc-disabled.a /opt/homebrew/Cellar/dotnet/9.0.101/libexec/packs/runtime.osx-arm64.Microsoft.DotNet.ILCompiler/9.0.0/sdk/libstdc++compat.a /opt/homebrew/Cellar/dotnet/9.0.101/libexec/packs/runtime.osx-arm64.Microsoft.DotNet.ILCompiler/9.0.0/framework/libSystem.Native.a /opt/homebrew/Cellar/dotnet/9.0.101/libexec/packs/runtime.osx-arm64.Microsoft.DotNet.ILCompiler/9.0.0/framework/libSystem.Globalization.Native.a /opt/homebrew/Cellar/dotnet/9.0.101/libexec/packs/runtime.osx-arm64.Microsoft.DotNet.ILCompiler/9.0.0/framework/libSystem.IO.Compression.Native.a /opt/homebrew/Cellar/dotnet/9.0.101/libexec/packs/runtime.osx-arm64.Microsoft.DotNet.ILCompiler/9.0.0/framework/libSystem.Net.Security.Native.a /opt/homebrew/Cellar/dotnet/9.0.101/libexec/packs/runtime.osx-arm64.Microsoft.DotNet.ILCompiler/9.0.0/framework/libSystem.Security.Cryptography.Native.Apple.a /opt/homebrew/Cellar/dotnet/9.0.101/libexec/packs/runtime.osx-arm64.Microsoft.DotNet.ILCompiler/9.0.0/framework/libSystem.Security.Cryptography.Native.OpenSsl.a --target=arm64-apple-macos12.0 -g -Wl,-rpath,'@executable_path' -lssl -lcrypto -lbrotlienc -lbrotlidec -ldl -lobjc -lswiftCore -lswiftFoundation -lz -licucore -lm -L/usr/lib/swift -framework CoreFoundation -framework CryptoKit -framework Foundation -framework Security -framework GSS" exited with code 1.

Build failed with 1 error(s) in 5.3s
```

If anyone knows what I can do to fix this, that'd be awesome!

**Comment by AhmedZero** [comment]
Is clang installed?

**Pull request merged**